02-peer-protocol.md

```diff
@@ -545,6 +545,42 @@ considered to be the channel funder (the sender of `open_channel`).
 The quiescence effect is exactly the same as if one had replied to the
 other.
 
+### `go_on`
+
+1. type: TODO (`go_on`)
+2. data:
+    * [`channel_id`:`channel_id`]
+    * **NOTE FOR REVIEWERS** do we need to include initiator here? I think not.
+
+#### Requirements
+
+The sender of `go_on`:
+  - MUST NOT send `go_on` for a channel that is NOT quiescent.
+  - MUST NOT send `go_on` if it is not the quiescence "initiator".
+  - MUST NOT send `go_on` twice without an interceding full quiescence
+    negotiation.
+  - MUST send `go_on` once any dependent protocol has completed.
+  - MAY send `update_*` messages afterwards.
+
+The receiver of `go_on`:
+  - if there is any incomplete downstream protocol state:
+    - SHOULD treat downstream protocol state as it would if the peer
+      disconnected.
```

Comment:
I think this would only make it worse with regards to the inner protocol. This indicates a buggy peer, so it would be safer to simply disconnect (which also guarantees the inner protocol will be aborted on both sides), wouldn't it?

```suggestion
    - SHOULD send a `warning` and disconnect.
```

Reply:
I see the interpretation as a buggy peer, however I see it differently. I see it as a forced abort of the inner protocol. I don't think things need to be buggy in order for you to behave that way. Indeed, the very fact that you are suggesting disconnecting as a means of handling this implies that the only sensible way to handle this partially executed inner protocol is the same.

So it depends on if we want to bless the process of aborting the inner protocol using this message or not. If you don't, then we can consider it a bug to arrive early and then it would make perfect sense to hang up. However, we could also just truncate the inner protocol state (which is what we do anyway) and then not have to reestablish the connection. I am sympathetic to using a disconnect here though as it may be the simplest path to resetting the inner protocol state and there may not be much to be gained by having this abort mechanism.

I think either approach is well-defined, and I think the particulars of how the implementation was done may make a "live truncation" of inner protocol state varying levels of difficult. If you think it drive you nuts to do this in Eclair we can just ban it and accept the suggestion you lay out here.

Reply:
> I think either approach is well-defined

I don't think that's correct. In this commit we're adding an explicit message to terminate the quiescence state to ensure that its state machine is well-defined independently of the inner protocols that may run once quiescent. By the same reasoning, every inner protocol must define its own clean termination mechanism as well.

In the case described in this requirement, the sender of `go_on` didn't use the correct inner protocol termination: we thus have no way of knowing how to make them terminate the inner protocol cleanly. The only general solution here is to disconnect, because if we don't, we can't know in what state they'll be?

Reply:
Again, I don't think the semantics are different. If you send an out-of-turn `go_on` it would have the same effect on the inner protocol as a disconnect. If you don't like it, there's no reason we have to go this route, but I think it *is* well-defined.

> we thus have no way of knowing how to make them terminate the inner protocol cleanly. The only general solution here is to disconnect, because if we don't, we can't know in what state they'll be?

Any inner protocol like this will have volatile state and committed state. The spec should delineate between the two as it is required for any disconnect tolerant protocol. As such, you know exactly what state they'll be in and can just revert any volatile state, which is exactly what you do during reconnect.

Sending an unexpected `go_on` always has a sensible interpretation. If you want to disconnect anyway, you can. You always can.

Does this deviate from Eclair's implementation strategy? I understand that it may be annoying to go this route and as such I'm willing to commit your suggestion, but I just want to make sure that we're on the same page that this is an effort/practicality discussion not a question of whether a solution exists.

Reply:
TL;DR: I'll add the language around "MAY disconnect" and that should unify concerns. If you want to make a case for why this is YAGNI, I'm open to it. I haven't thought about the strength-to-weight ratio of such a feature. I was just trying to write as permissive of a spec as possible.

---

> The receiver cannot know in which state the sender is

The receiver only needs to know that the protocol has been aborted.

>but it has no guarantee that the sending node did the same

Good catch. I'll admit I should probably add the same requirement to the sender to truncate volatile state to be clear. 

> it only knows that it's buggy so it could be in any state

Buggy implies that doing a forced abort can only result from incorrect implementation, and that is not a valid assumption in context right now. If you want to make the claim that we should *never* do a forced abort of some negotiation and that any protocol that supports an abort operation MUST specify one, then we can. However, I'm aware of no such requirement up until this point.

I think we have to be a little more careful about what behavior we classify as "buggy". Given that the spec language I propose explicitly allows it, it -- by definition -- would not be buggy to send those messages out-of-turn.

I'm operating with the understanding that a bug is when the software either doesn't implement the spec, or if the spec violates an assumed property. In this case, I'm not sure what assumed property would be violated.

Because any property that would be violated by an out-of-turn `go_on` would necessarily be violated by a disconnect in the same place, and since disconnects are assumed to be *always allowed*, there must exist well-defined behavior for this case and if we want to take advantage of that well-defined behavior in the future, we shouldn't classify it as "buggy". We may deem that the implementation effort of that well-defined behavior isn't worth what it buys us. Fair enough! However, it's only buggy if we decide to define it as buggy, which is currently what's under debate!

Beyond that, I consider it a layer violation to make the validity of the messages defined in quiescence *depend on* the in-this-scope-unknowable state of the downstream protocol. So I would consider it incoherent to put the following statement into the spec:

> MUST NOT send `go_on` while a downstream protocol session is active.

...or something to that effect. This is because while quiescence is aware that there may exist downstream protocols, it shouldn't be allowed to know any of the particulars of what they are and how they work. I contrast this with the current language where we make a recommendation for how the downstream protocol handles this case since the awareness of protocol state flows from dependency to dependent *AND NOT* the other way around.

> As a rule of thumb, whenever we detect buggy behavior from our peer, I believe we should disconnect, because we have no way of knowing how messed up their current state is, and restarting "from scratch" is our best bet at fixing it, since we have mechanisms in channel_reestablish to get back to the latest safe point. 

I wholeheartedly agree with this, modulo my philosophy on "buggy" behavior from above.